cardano-submit-api/src/Cardano/TxSubmit/Web.hs

```diff
@@ -99,32 +102,32 @@ readEnvSocketPath =
     envName :: String
     envName = "CARDANO_NODE_SOCKET_PATH"
 
-readByteStringTextEnvelopeAnyOf2
-  :: [FromSomeType HasTextEnvelope b]
+deserialiseOne :: forall b. ()
+  => FromSomeType SerialiseAsCBOR b
   -> ByteString
-  -> ExceptT TextEnvelopeError IO b
-readByteStringTextEnvelopeAnyOf2 types content = hoistEither $ do
-  te <- first TextEnvelopeAesonDecodeError $ Aeson.eitherDecodeStrict' content
-  deserialiseFromTextEnvelopeAnyOf types te
-
-readByteStringInAnyCardanoEra
-  :: ( HasTextEnvelope (thing ByronEra)
-     , HasTextEnvelope (thing ShelleyEra)
-     , HasTextEnvelope (thing AllegraEra)
-     , HasTextEnvelope (thing MaryEra)
-     )
-  => (forall era. AsType era -> AsType (thing era))
+  -> Either DecoderError b
+deserialiseOne (FromSomeType ttoken f) bs = f <$> deserialiseFromCBOR ttoken bs
+
+deserialiseAnyOf :: forall b. ()
+  => [FromSomeType SerialiseAsCBOR b]
   -> ByteString
-  -> ExceptT TextEnvelopeError IO (InAnyCardanoEra thing)
-readByteStringInAnyCardanoEra asThing = readByteStringTextEnvelopeAnyOf2
-  [ FromSomeType (asThing AsByronEra)   (InAnyCardanoEra ByronEra)
-  , FromSomeType (asThing AsShelleyEra) (InAnyCardanoEra ShelleyEra)
-  , FromSomeType (asThing AsAllegraEra) (InAnyCardanoEra AllegraEra)
-  , FromSomeType (asThing AsMaryEra)    (InAnyCardanoEra MaryEra)
-  ]
+  -> Either RawCborDecodeError b
+deserialiseAnyOf ts te = let (es, as) = partitionEithers results in maybe (errors es) Right (listToMaybe as)
+  where
```

Comment:
An alternative suggestion: https://github.com/input-output-hk/cardano-node/commit/2944c3845e1f8003675e02a055ec3a6dcac9aa27

Reply:
I've take your suggestion to use list instead of non-empty-list, but also fixed up the logic like this:  If there is at least successful decodes, always take the first one.